File: LeetCode/Medium/1545-find-kth-bit-in-nth-binary-string/1545-find-kth-bit-in-nth-binary-string.py
class Solution:
    def findKthBit(self, n: int, k: int) -> str:
        def dfs(n, k):
            if n == 1:
                return "0"
            
            length = 2 ** n - 1
            mid = 2 ** (n - 1)

            if k == mid:
                return "1"
            elif k < mid:
                return dfs(n-1, k)
            else:
                return "1" if dfs(n-1, length - k + 1) == "0" else "0"

        return dfs(n, k)

File: LeetCode/Medium/1545-find-kth-bit-in-nth-binary-string/1545-find-kth-bit-in-nth-binary-string.java
class Solution {
    public char findKthBit(int n, int k) {
        return dfs(n, k);
    }
    
    private char dfs(int n, int k){
        if (n == 1) return '0';
        
        int len = (int)Math.pow(2, n) - 1;
        int mid = (int)Math.pow(2, n-1);
        
        if (k == mid) return '1';
        else if (k < mid) return dfs(n-1, k);
        else return dfs(n-1, len - k + 1) == '1' ? '0' : '1';
    }
}